File: backend/app/schemas/product_catalog.py
"""Phase 18.2: request/response schemas for the platform product catalog.

Mirrors the style of app/schemas/plan.py -- pure serialization/validation
shapes, no business logic. ORM objects are never returned directly.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class ProductSuiteCreateRequest(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    display_order: int = 0
    is_active: bool = True


class ProductSuiteUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    display_order: int | None = None
    is_active: bool | None = None


class ProductSuiteResponse(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    description: str | None
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProductModuleCreateRequest(BaseModel):
    suite_id: uuid.UUID
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    display_order: int = 0
    is_active: bool = True


class ProductModuleUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    display_order: int | None = None
    is_active: bool | None = None


class ProductModuleResponse(BaseModel):
    id: uuid.UUID
    suite_id: uuid.UUID
    code: str
    name: str
    description: str | None
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProductPageCreateRequest(BaseModel):
    module_id: uuid.UUID
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    route: str | None = Field(default=None, max_length=255)
    display_order: int = 0
    is_active: bool = True


class ProductPageUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    route: str | None = Field(default=None, max_length=255)
    display_order: int | None = None
    is_active: bool | None = None


class ProductPageResponse(BaseModel):
    id: uuid.UUID
    module_id: uuid.UUID
    code: str
    name: str
    description: str | None
    route: str | None
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProductFeatureCreateRequest(BaseModel):
    module_id: uuid.UUID
    code: str = Field(min_length=1, max_length=128)
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    is_active: bool = True


class ProductFeatureUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    is_active: bool | None = None


class ProductFeatureResponse(BaseModel):
    id: uuid.UUID
    module_id: uuid.UUID
    code: str
    name: str
    description: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ProductModuleWithChildrenResponse(ProductModuleResponse):
    pages: list[ProductPageResponse] = []
    features: list[ProductFeatureResponse] = []


class ProductSuiteWithChildrenResponse(ProductSuiteResponse):
    modules: list[ProductModuleWithChildrenResponse] = []

File: frontend/lib/api/productCatalog.ts
// Phase 18.2: typed REAL-mode client for platform product catalog
// administration (backend/app/api/v1/product_catalog.py). Requires
// PLATFORM_MANAGE on the backend — a non-platform-admin user calling these
// gets a 403 regardless of what the frontend shows. This client is
// intentionally thin, following lib/api/plan.ts's exact pattern: it only
// maps HTTP requests/responses to typed shapes. All validation/audit/
// transaction behavior lives in app.services.product_catalog_service.

import { apiRequest } from "@/lib/apiClient";

export interface ProductSuiteResponse {
  id: string;
  code: string;
  name: string;
  description: string | null;
  display_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProductModuleResponse {
  id: string;
  suite_id: string;
  code: string;
  name: string;
  description: string | null;
  display_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProductPageResponse {
  id: string;
  module_id: string;
  code: string;
  name: string;
  description: string | null;
  route: string | null;
  display_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProductFeatureResponse {
  id: string;
  module_id: string;
  code: string;
  name: string;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProductModuleWithChildrenResponse extends ProductModuleResponse {
  pages: ProductPageResponse[];
  features: ProductFeatureResponse[];
}

export interface ProductSuiteWithChildrenResponse extends ProductSuiteResponse {
  modules: ProductModuleWithChildrenResponse[];
}

export interface ProductSuiteCreateRequest {
  code: string;
  name: string;
  description?: string | null;
  display_order?: number;
  is_active?: boolean;
}

export interface ProductModuleCreateRequest {
  suite_id: string;
  code: string;
  name: string;
  description?: string | null;
  display_order?: number;
  is_active?: boolean;
}

export interface ProductPageCreateRequest {
  module_id: string;
  code: string;
  name: string;
  description?: string | null;
  route?: string | null;
  display_order?: number;
  is_active?: boolean;
}

export interface ProductFeatureCreateRequest {
  module_id: string;
  code: string;
  name: string;
  description?: string | null;
  is_active?: boolean;
}

export const productCatalogApi = {
  getCatalogTree: (accessToken: string) =>
    apiRequest<ProductSuiteWithChildrenResponse[]>("/platform/product-catalog", { accessToken }),

  listSuites: (accessToken: string) =>
    apiRequest<ProductSuiteResponse[]>("/platform/product-suites", { accessToken }),

  createSuite: (accessToken: string, payload: ProductSuiteCreateRequest) =>
    apiRequest<ProductSuiteResponse>("/platform/product-suites", { method: "POST", body: payload, accessToken }),

  createModule: (accessToken: string, payload: ProductModuleCreateRequest) =>
    apiRequest<ProductModuleResponse>("/platform/product-modules", { method: "POST", body: payload, accessToken }),

  createPage: (accessToken: string, payload: ProductPageCreateRequest) =>
    apiRequest<ProductPageResponse>("/platform/product-pages", { method: "POST", body: payload, accessToken }),

  createFeature: (accessToken: string, payload: ProductFeatureCreateRequest) =>
    apiRequest<ProductFeatureResponse>("/platform/product-features", {
      method: "POST",
      body: payload,
      accessToken,
    }),
};
